main.go

```diff
@@ -117,30 +122,134 @@ type ObjRef struct {
+// It updates storetheindex with the new CIDs, saying they are present on autoretrieve
 // With that, clients using bitswap can query autoretrieve servers using bitswap and get data from estuary
 func (s *Server) updateAutoretrieveIndex(tickInterval time.Duration, quit chan struct{}) error {
-	var autoretrieves []Autoretrieve
+	var autoretrieves []autoretrieve.Autoretrieve
 	var lastTickTime time.Time
-	ticker := time.NewTicker(tickInterval)
+	var curTime time.Time
+	var newContextID []byte
+	var err error
+
+	// Create index-provider engine (s.Node.IndexProvider) to send announcements to
+	// this needs to keep running continuously because storetheindex
+	// will come to fetch for advertisements "when it feels like it"
+	s.Node.ArEngine, err = autoretrieve.NewAutoretrieveEngine()
+	if err != nil {
+		return err
+	}
 
+	s.Node.ArEngine.RegisterMultihashLister(func(ctx context.Context, contextID []byte) (provider.MultihashIterator, error) {
+		var newContents []Content
+		var ar autoretrieve.Autoretrieve
+
+		arHandle, err := autoretrieve.ParseContextID(contextID)
+		if err != nil {
+			return nil, err
+		}
+
+		// get the autoretrieve entry from the database
+		err = s.DB.Find(&ar, "handle = ?", arHandle).Error
+		if err != nil {
+			return nil, err
+		}
+
+		// find all new multihashes since the last time we advertised for this autoretrieve server
+		err = s.DB.Find(&newContents, "active = true and created_at >= ?", ar.LastAdvertisement).Error
+		if err != nil {
+			return nil, err
+		}
+
+		multihashes := []multihash.Multihash{}
+		for _, content := range newContents {
+			multihashes = append(multihashes, content.Cid.CID.Hash())
+		}
+
+		return &autoretrieve.SimpleEstuaryMhIterator{
+			Mh: multihashes,
+		}, nil
+	})
+
+	// start engine
+	s.Node.ArEngine.Start(context.Background())
+	defer s.Node.ArEngine.Shutdown()
+
+	// start ticker
+	ticker := time.NewTicker(tickInterval)
 	defer ticker.Stop()
-	for {
-		lastTickTime = time.Now().UTC().Add(-tickInterval)
 
+	for {
+		curTime = time.Now()
+		lastTickTime = curTime.Add(-tickInterval)
 		// Find all autoretrieve servers that are online (that sent heartbeat)
-		err := s.DB.Find(&autoretrieves, "last_connection > ?", lastTickTime).Error
+		err = s.DB.Find(&autoretrieves, "last_connection > ?", lastTickTime).Error
 		if err != nil {
 			log.Errorf("unable to query autoretrieve servers from database: %s", err)
 			return err
 		}
-		if len(autoretrieves) > 0 {
-			for _, ar := range autoretrieves {
-				fmt.Println("online: ", ar) // TODO: remove
+		if len(autoretrieves) == 0 {
+			log.Infof("no autoretrieve servers online")
+			// wait for next tick, or quit
+			select {
+			case <-ticker.C:
+				continue
+			case <-quit:
+				break
 			}
-		} else {
-			log.Info("no autoretrieve servers online")
+		}
+
+		log.Infof("announcing new CIDs to %d autoretrieve servers", len(autoretrieves))
+		// send announcement with new CIDs for each autoretrieve server
+		for _, ar := range autoretrieves {
```

Comment:
Also good to move this loop into a function, and call the function in a goroutine at each tick